subtrack.MAUI/Services/MonthlyPageCalculator.cs

```diff
@@ -0,0 +1,45 @@
+﻿using subtrack.DAL.Entities;
+using subtrack.MAUI.Responses;
+using subtrack.MAUI.Services.Abstractions;
+
+namespace subtrack.MAUI.Services
+{
+    internal class MonthlyPageCalculator : IMonthlyPageCalculator
+    {
+        private readonly ISubscriptionsCalculator subscriptionsCalculator;
+        private readonly ISubscriptionService subscriptionService;
+
+        public MonthlyPageCalculator(ISubscriptionsCalculator subscriptionsCalculator, ISubscriptionService subscriptionService)
+        {
+            this.subscriptionsCalculator = subscriptionsCalculator;
+            this.subscriptionService = subscriptionService;
+        }
+
+        private IEnumerable<Subscription> GetPaymentsUntilMonth(Subscription subscription, DateTime fromIncludedDate, DateTime toIncludedDate)
+        {
+            subscription.LastPayment = subscriptionsCalculator.GetNextPaymentDate(subscription);
+
+            while (subscription.LastPayment.Date >= fromIncludedDate.Date
+                    && subscription.LastPayment.Date <= toIncludedDate.Date)
+            {
+                yield return (Subscription)subscription.Clone();
+                subscription.LastPayment = subscriptionsCalculator.GetNextPaymentDate(subscription);
+            }
+        }
+
+        public async Task<IEnumerable<IGrouping<int, SubscriptionsMonthResponse>>> GetMonthlySubscriptionLists(DateTime fromIncludedMonthDate, DateTime finalIncludedMonthDate)
+        {
+            var subscriptions = await subscriptionService.GetAllAsync();
```

Comment:
This an unnecessary dependency

Reply:
please elaborate.

Reply:
Instead of having a dependency on isub we can let the caller pass in the subs. That way we also dont force a reload of subs in case caller already has the subs. 

Input/output calls should always be centralized and the caller should be responsible for doing them

Reply:
We have, and will have only one caller for this service.

Reply:
Unnecessary risk to take. Cant be sure of that and it even makes testing worse as we have to mock the subscriptionservice

Reply:
We'll refactor if it comes to that. No point in anticipating future needs that may or may not materialize.

Reply:
> We'll refactor if it comes to that. No point in anticipating future needs that may or may not materialize.

True but its a highly likely case. Limiting your caller, forcing them to refactor when you vould have just left a simple parameter in the params doesnt make sense, why would you consider adding a dependency is good design? I dont see any benefits of doing so unless it simplified something for the caller, which it does not. It just adds magic

Reply:
I disagree but I'll make the change if you feel this strongly about it.

Reply:
You havent told me why you think its a good idea to add the dependency

Reply:
At the moment, we have only one caller (the one we refactored for), and as far as the caller is concerned, it needs a list of subscriptions grouped by a certain criteria. This makes the interface a lot simpler. 

Ideally I would have also made it so that all the caller needs to pass the number of years of future data which is needed (2 in our case) and the service would handle the rest. 

This of course, assumes that the service would be used exclusively by this caller so it does call for a refactor if/when that changes.

Reply:
I've run into a major issue, and will need you to make this change

 ![IMG_20230822_120258969.jpg](https://github.com/Code2Gether-Discord/subtrack/assets/12110372/22453573-476e-435b-be10-2163745982dc)

Reply:
I can understand you see it that way, but the interface doesnt really get simpler - it gets more complicated as the app grows because the subs are magically fetched by this method. If we centralized the fetching then they can be reused for example for other calculations (which is not entirely impossible for this view, maybe next to year we decide to show the total cost for only that year) and we could do so without having to modify the source method - which allows for Open/Closed principle in SOLID.

Restricting input I can understand but at the same time it does not add any value of doing so. But not doing so does add value because we allow them to get from a specific month to a specific month and thats another not so impossible case we might have. Instead of blindly saying you can see 2 years in the future, maybe we say you can see untill the last sub is due f.e if you have yearly every 3 years or whatever.

Adding that line of code doesnt simplify anything. Its only one line of code to fetch subs. If it was something complicated then ok I would agree with you but there is no business logic involved when fetching a sub. 

It does make things harder, limits the caller to reuse subs, limits them to filter, forces them to provide ISubService, forces our tests to mock it and the caller has no idea where we fetch the subs from without having to pay us a visit

Reply:
> I've run into a major issue, and will need you to make this change
> 
>  ![IMG_20230822_120258969.jpg](https://github.com/Code2Gether-Discord/subtrack/assets/12110372/22453573-476e-435b-be10-2163745982dc)
> 
> 

Rip lol. Ill do it after work